src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -407,105 +407,96 @@ internal void NotifyActivityAddException(Activity activity, Exception exception,
         }
     }
 
-    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
-    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
-    // during the enumeration which force restarting the enumeration again.
-    // Caution: We can have the action executed on the same item more than once which is ok in our scenarios.
+    //this class uses Interlocked operations and a copy-on-write design to ensure thread safety
+    //all operations are thread safe
     internal sealed class SynchronizedList<T>
     {
-        private readonly List<T> _list;
-        private uint _version;
-
-        public SynchronizedList() => _list = new List<T>();
+        //This array must not be written to directly. Copy the array and then replace it with the new array.
+        private T[] _volatileArray;
+        public SynchronizedList() => _volatileArray = [];
 
         public void Add(T item)
         {
-            lock (_list)
+            while (true)
             {
-                _list.Add(item);
-                _version++;
+                T[] local = _volatileArray;
+                var newArray = new T[local.Length + 1];
+                Array.Copy(local, newArray, local.Length);
+                newArray[local.Length] = item;
+
+                if (Interlocked.CompareExchange(ref _volatileArray, newArray, local) == local)
+                    break;
             }
         }
 
         public bool AddIfNotExist(T item)
         {
-            lock (_list)
+            while (true)
             {
-                if (!_list.Contains(item))
+                T[] local = _volatileArray;
+
+                foreach (T arrayItem in local)
                 {
-                    _list.Add(item);
-                    _version++;
-                    return true;
+                    if (EqualityComparer<T>.Default.Equals(arrayItem, item))
```

Comment:
You can use `Array.IndexOf`. It is what the existing `List.Contains` call uses underneath.

Reply:
Fixed